### PR TITLE
KtBridge: call Kotlin/JS from Kotlin/JVM

### PR DESCRIPTION
**Pull request by swankjesse**
This is the first step in a more ambitious project.

### CODE REVIEW

**Comment by swankjesse** [comment]
Next steps:
 * [x] Generate Java so we don't need a dynamic proxy
 * [x] Generate JS so we don't need to use `@JsName` and so that functions can be overloaded
 * [ ] Integrate either kotlinx.serialization or Wire for a built-in JsAdapter
 * [ ] Support suspend functions
 * [ ] Bridge from JVM back to JS

File: build.gradle [hunk 1]
```diff
@@ -5,11 +5,15 @@ buildscript {
     mavenCentral()
     google()
     gradlePluginPortal()
+    maven {
+      url "https://plugins.gradle.org/m2/"
```
Comment (JakeWharton) [inline build.gradle]: We already have this one line up

File: ktbridge/build.gradle [hunk 2]
```diff
@@ -0,0 +1,40 @@
+apply plugin: 'org.jetbrains.kotlin.multiplatform'
+apply plugin: 'com.vanniktech.maven.publish'
+
+kotlin {
+  jvm {
+    withJava()
+  }
```
Comment (JakeWharton) [inline ktbridge/build.gradle:7]:
```suggestion
  jvm()
```

File: ktbridge/build.gradle [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+  js {
+    browser {
+    }
+  }
+
+  sourceSets {
+    commonMain {
+      dependencies {
+        implementation deps.okioMultiplatform
```
Comment (JakeWharton) [inline ktbridge/build.gradle]:
```suggestion
        api deps.okioMultiplatform
```

File: ktbridge/testing/build.gradle [hunk 4]
```diff
@@ -0,0 +1,33 @@
+apply plugin: 'org.jetbrains.kotlin.multiplatform'
+
+kotlin {
+  jvm {
+    withJava()
+  }
```
Comment (JakeWharton) [inline ktbridge/testing/build.gradle]:
```suggestion
  jvm()
```

File: ktbridge/testing/build.gradle [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+  js {
+    browser {
+    }
+  }
+
+  sourceSets {
+    commonMain {
+      dependencies {
+        implementation deps.okioMultiplatform
```
Comment (JakeWharton) [inline ktbridge/testing/build.gradle]:
```suggestion
```
Not needed after the switch to `api` above

File: ktbridge/testing/build.gradle [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+        implementation project(":ktbridge")
+      }
+    }
+    jsMain {
+      dependencies {
+        implementation project(":ktbridge")
+      }
+    }
```
Comment (JakeWharton) [inline ktbridge/testing/build.gradle]:
```suggestion
```
commonMain dep will be available to all platforms

**Review by JakeWharton — COMMENTED** [review]
Too hard to review on mobile

File: build.gradle (hunk 1 above)
Comment (swankjesse) [reply build.gradle]: Ooooh ouch. Fixed.

File: ktbridge/build.gradle (hunk 2 above)
Comment (swankjesse) [reply ktbridge/build.gradle:7]: Fixed.

File: ktbridge/build.gradle (hunk 3 above)
Comment (swankjesse) [reply ktbridge/build.gradle]: Fixed!

File: ktbridge/testing/build.gradle (hunk 4 above)
Comment (swankjesse) [reply ktbridge/testing/build.gradle]: Fixed.

File: ktbridge/testing/build.gradle (hunk 5 above)
Comment (swankjesse) [reply ktbridge/testing/build.gradle]: Fixed.

File: ktbridge/testing/build.gradle (hunk 6 above)
Comment (swankjesse) [reply ktbridge/testing/build.gradle]: Nice, yes of course. Fixed.

File: ktbridge/build.gradle (hunk 2 above)
Comment (swankjesse) [reply ktbridge/build.gradle:7]: Hmmm... that didn’t work.

```
./gradlew clean ktbridge:tasks

> Task :ktbridge:tasks FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':ktbridge:tasks'.
> Could not create task ':ktbridge:javadocsJar'.
   > Could not create task of type 'JavadocsJar'.
      > Task with name 'javadoc' not found in project ':ktbridge'.
```

File: .github/workflows/release.yaml [hunk 7]
```diff
@@ -42,6 +42,9 @@ jobs:
 
       - run: ./gradlew :quickjs:jvm:build
 
+      - name: KtBridge
+        run: ./gradlew :ktbridge:check
```
Comment (swankjesse) [inline .github/workflows/release.yaml:46]: I added this. I don’t have opinions on how these are organized so if you’d prefer a separate job that works for me.

**Comment by JakeWharton** [comment]
Just FYI I didn't actually review anything but the build files yet.

**Comment by swankjesse** [comment]
Sounds good. Lots of this is bootstrapping, and I expect the API to change significantly before anything safe to release. Also I'm open to splitting this off into another project if you'd prefer that.

**Comment by JakeWharton** [comment]
(Still having not reviewed...)

I'm tempted to rewrite the existing Java into Kotlin and move this into the core. Make it one of the core value propositions. It also opens up the ability to support `suspend` functions there as well as types like UByteArray, UInt, etc.

**Comment by swankjesse** [comment]
One more thing that'd unlock: we could expand this project to integrate QuickJS on iOS via Kotlin/Native. Seems like a big win to me!

File: ktbridge/build.gradle (hunk 2 above)
Comment (JakeWharton) [reply ktbridge/build.gradle:7]: Need to apply the Dokka plugin since this is Kotlin

**Pull request merged**